cloudinit/stages.py

```diff
@@ -692,34 +702,81 @@ def _find_networking_config(self):
+
+        for evt_scope in scopes:
+            if event_source_type in allowed.get(evt_scope, []):
+                LOG.debug('Event Allowed: scope=%s EventType=%s',
+                          evt_scope, event_source_type)
+                return True
+
+        LOG.debug('Event Denied: scopes=%s EventType=%s',
+                  scopes, event_source_type)
+        return False
+
     def _apply_netcfg_names(self, netcfg):
         try:
             LOG.debug("applying net config names for %s", netcfg)
             self.distro.apply_network_config_names(netcfg)
         except Exception as e:
             LOG.warning("Failed to rename devices: %s", e)
 
+    def _is_first_boot(self):
+        bpath = os.path.join(
+            self.paths.get_cpath(),
+            'sem',
+            'config_scripts_per_once.once'
+        )
+        first_boot = not os.path.exists(bpath)
+        LOG.debug('Is first boot? %s', first_boot)
+        return first_boot
+
     def apply_network_config(self, bring_up):
-        # get a network config
+        """Apply the network config.
+
+        Find the config, determine whether to apply it, apply it via
+        the distro, and optionally bring it up
+        """
+
         netcfg, src = self._find_networking_config()
         if netcfg is None:
             LOG.info("network config is disabled by %s", src)
             return
 
-        # request an update if needed/available
-        if self.datasource is not NULL_DATA_SOURCE:
-            if not self.is_new_instance():
-                if not self.datasource.update_metadata([EventType.BOOT]):
-                    LOG.debug(
-                        "No network config applied. Neither a new instance"
-                        " nor datasource network update on '%s' event",
-                        EventType.BOOT)
-                    # nothing new, but ensure proper names
-                    self._apply_netcfg_names(netcfg)
-                    return
-                else:
-                    # refresh netcfg after update
-                    netcfg, src = self._find_networking_config()
+        current_boot_type = EventType.BOOT
+        if self.is_new_instance():
+            current_boot_type = EventType.BOOT_NEW_INSTANCE
+
+        if not any([
+            self.datasource is NULL_DATA_SOURCE,
+            self.is_new_instance(),  # or should it be self._is_first_boot(),
+            (self.update_event_enabled(current_boot_type, scope='network')
+             and self.datasource.update_metadata([current_boot_type]))
+        ]):
+            LOG.debug(
+                "No network config applied. Neither a new instance"
+                " nor datasource network update on '%s' event",
+                EventType.BOOT)
+            # nothing new, but ensure proper names
+            self._apply_netcfg_names(netcfg)
+            return
+        else:
+            # refresh netcfg after update
+            netcfg, src = self._find_networking_config()
```

Comment:
A nit, for sure, but as we unconditionally return in the `if` branch, we don't need the `else`:

```suggestion
        # refresh netcfg after update
        netcfg, src = self._find_networking_config()
```